packages/vite/src/node/ssr/ssrModuleLoader.ts

```diff
@@ -178,62 +182,36 @@ async function instantiateModule(
   return Object.freeze(ssrModule)
 }
 
-function nodeRequire(
+// In node@12+ we can use dynamic import to load CJS and ESM
+// Caveat https://github.com/vitejs/vite/pull/5197#discussion_r722583544
+async function nodeImport(
   id: string,
   importer: string | null,
-  root: string,
-  preserveSymlinks: boolean
+  config: ViteDevServer['config'],
 ) {
-  const mdl = require(resolve(id, importer, root, preserveSymlinks));
-  return proxyESM(mdl);
+  const url = pathToFileURL(resolve(id, importer, config.root, !!config.resolve.preserveSymlinks))
+  const mod = await nativeImport(url);
+  return proxyESM(mod);
 }
 
-function nodeRequireOrImport(
+async function handleImport(
   id: string,
   importer: string | null,
-  root: string,
-  preserveSymlinks: boolean,
-  server: ViteDevServer,
+  config: ViteDevServer['config'],
 ): Promise<any> {
-  return new Promise((res, reject) => {
-    if (id.startsWith('node:')) {
-      return res(NATIVE_IMPORT(id.slice('node:'.length)))
-    }
-    try {
-      const mdl = nodeRequire(id, importer, root, preserveSymlinks);
-      return res(mdl);
-    } catch (e) {
-      const resolvedPath = resolve(id, importer, root, preserveSymlinks);
-      if (e instanceof SyntaxError && /export|import/.test(e.message)) {
-        server.config.logger.error(
-          `Failed to load "${id}"!\nESM format is not natively supported in "node@${process.version}".\nPlease use CommonJS or upgrade to an LTS version of node above "node@12.17.0".`,
-        );
-      } else if (e.code === 'ERR_REQUIRE_ESM') {
-        const url = pathToFileURL(resolvedPath);
-        try {
-          return NATIVE_IMPORT(url.toString()).then(
-            (mdl: any) => {
-              const mod = proxyESM(mdl, true);
-              return res(mod);
-            },
-            (err: any) => {
-              reject(err)
-            }
-        )
-        } catch (e) {
-          reject(e);
-        }
-      }
+    // `resolve` doesn't handle `node:` builtins, so handle them directly
+    if (id.startsWith('node:') || builtins.has(id)) {
+      return nativeImport(id);
     }
-  })
+    return nodeImport(id, importer, config);
 }
 
 // rollup-style default import interop for cjs
-function proxyESM(mod: any, isESM = mod.__esModule) {
-  const defaultExport = isESM ? mod.default : mod
+function proxyESM(mod: any) {
+  const isESM = !!(mod.__esModule ?? mod.default);
   return new Proxy(mod, {
     get(mod, prop) {
-      if (prop === 'default') return defaultExport
+      if (prop === 'default') return isESM ? mod.default : mod;
```

Comment:
IIUC, this change allows for `import M from 'm'` to mimic `import * as M from 'm'` when the `m` module is ESM but has no `export default` declaration. Is that right?

Reply:
That's correct.

FWIW this isn't a new change, you can see [the current code](https://github.com/vitejs/vite/blob/8fe8df37d4b68705bfb2f768b47ca99a8678f4e9/packages/vite/src/node/ssr/ssrModuleLoader.ts#L184-L191). This diff is just because it's slightly refactored.